File: python/ch08_agent/agent_ch08/features/todos.py
"""第六章子 Agent 会话级 TODO 计划快照。

这是什么：管理 Agent 会话内 TODO 列表的模块
Java 类比：类似有状态的 TodoService，实现 ToolRoundObserver 接口
为什么需要：让 Agent 维护任务清单，并在长期未更新时提醒模型同步计划

Java 对照：`TodoTracker` 同时扮演一个有状态的领域服务和工具轮观察器。
它不是数据库 Repository，状态只属于当前 AgentRunner，会话结束后自然消失。
"""

import json
from collections.abc import Mapping, Sequence
from dataclasses import dataclass
from typing import Any, Literal

from ..core.messages import ChatMessage, system_message
from ..core.tools import ToolContext, ToolDefinition, ToolResult, tool_success

MAX_TODOS = 50
STALE_TOOL_ROUNDS = 3
TODO_STALE_REMINDER = (
    "请保持 TODO 列表为最新状态。计划发生变化时，请调用 todo_write 提交完整任务快照。"
)
TodoStatus = Literal["pending", "in_progress", "completed"]
TODO_STATUSES: tuple[TodoStatus, ...] = ("pending", "in_progress", "completed")


@dataclass(frozen=True, slots=True)
class TodoItem:
    """一条不可变 TODO，类似 Java record。

    这是什么：表示单个 TODO 条目的数据类
    Java 类比：类似 record TodoItem(String content, TodoStatus status)
    为什么需要：以不可变方式存储任务内容和状态，防止意外修改
    """

    content: str  # 去掉首尾空白后的任务说明。
    status: TodoStatus  # pending、in_progress、completed 三选一。


def _validate_todo_input(value: Mapping[str, Any]) -> bool:
    """严格校验完整快照，未知字段或任意坏项都整体拒绝。

    这是什么：校验 todo_write 工具参数格式的函数
    Java 类比：类似 boolean validateTodoInput(Map<String, Object> args)
    为什么需要：确保 TODO 快照格式正确，防止无效数据进入系统
    """
    if set(value) != {"todos"}:
        return False
    raw_todos = value.get("todos")
    if not isinstance(raw_todos, list) or len(raw_todos) > MAX_TODOS:
        return False
    for item in raw_todos:
        if not isinstance(item, dict) or set(item) != {"content", "status"}:
            return False
        content = item.get("content")
        status = item.get("status")
        if not isinstance(content, str) or not content.strip() or status not in TODO_STATUSES:
            return False
    return True


def _serialize_snapshot(todos: Sequence[TodoItem]) -> str:
    """返回紧凑、字段顺序稳定的 ASCII JSON，中文会变成 Unicode 转义。

    这是什么：将 TODO 列表序列化为 JSON 字符串的函数
    Java 类比：类似 String serializeToJson(List<TodoItem> todos)
    为什么需要：生成稳定格式的 JSON 响应，便于模型解析和测试断言
    """
    return json.dumps(
        {"todos": [{"content": item.content, "status": item.status} for item in todos]},
        ensure_ascii=True,
        separators=(",", ":"),
    )


class TodoTracker:
    """保存当前完整计划，并统计连续未更新计划的工具轮。

    这是什么：管理 TODO 列表状态并监控更新频率的类
    Java 类比：类似 @Service class TodoTracker implements ToolRoundObserver
    为什么需要：提供 TODO 工具实现，同时监控未更新轮数并在必要时提醒模型
    """

    def __init__(self) -> None:
        self._todos: tuple[TodoItem, ...] = ()
        self._non_todo_tool_rounds = 0
        self.tool_definition = ToolDefinition(
            "todo_write",
            "用完整任务快照替换当前 TODO 列表；计划变化时需要重新提交全部条目。",
            {
                "type": "object",
                "properties": {
                    "todos": {
                        "type": "array",
                        "maxItems": MAX_TODOS,
                        "items": {
                            "type": "object",
                            "properties": {
                                "content": {"type": "string", "minLength": 1},
                                "status": {"type": "string", "enum": list(TODO_STATUSES)},
                            },
                            "required": ["content", "status"],
                            "additionalProperties": False,
                        },
                    }
                },
                "required": ["todos"],
                "additionalProperties": False,
            },
            # 这里的 write 表示“修改会话状态”，不是写磁盘。第三章审批规则只匹配
            # write_file/edit_file，因此 todo_write 不会错误地弹出文件审批。
            "write",
            self._write_todos,
            _validate_todo_input,
        )

    @property
    def todos(self) -> tuple[TodoItem, ...]:
        """返回当前不可变快照，外部只能读取。"""
        return self._todos

    def _write_todos(self, arguments: Mapping[str, Any], _context: ToolContext) -> ToolResult:
        """一次性替换全部计划；进入此方法前参数已经完整校验。"""
        raw_todos = arguments["todos"]
        if not isinstance(raw_todos, tuple):
            raise TypeError("todo_write 参数在 prepare 后必须冻结为 tuple")
        self._todos = tuple(
            TodoItem(str(item["content"]).strip(), item["status"]) for item in raw_todos
        )
        self._non_todo_tool_rounds = 0
        return tool_success(_serialize_snapshot(self._todos))

    def record_tool_round(self, tool_names: Sequence[str]) -> None:
        """按整轮而不是按调用数统计；一轮出现 todo_write 就重置。"""
        if not tool_names:
            return
        if self.tool_definition.name in tool_names:
            self._non_todo_tool_rounds = 0
            return
        self._non_todo_tool_rounds += 1

    def before_model(self) -> tuple[ChatMessage, ...]:
        """达到三轮后只为下一次请求返回一条临时系统提醒。"""
        if self._non_todo_tool_rounds < STALE_TOOL_ROUNDS:
            return ()
        self._non_todo_tool_rounds = 0
        return (system_message(TODO_STALE_REMINDER),)

File: typescript/chapters/ch07/src/features/todos.ts
import { z } from "zod";

import type { ChatMessage } from "../core/messages.js";
import { systemMessage } from "../core/messages.js";
import type { ToolDefinition, ToolResult } from "../core/tools.js";
import { toolSuccess } from "../core/tools.js";

// 单次计划快照的数量上限，防止模型一次性塞入超大数组。
export const MAX_TODOS = 50;
// 连续未更新计划的其他工具轮次数，达到后会在下一次模型请求前提醒。
export const STALE_TOOL_ROUNDS = 3;
export const TODO_STALE_REMINDER =
  "Keep the TODO list current. Call todo_write with the complete task snapshot when the plan changes.";
// 三个合法状态作为工具 schema 的唯一取值来源。
export const TODO_STATUSES = Object.freeze(["pending", "in_progress", "completed"] as const);
// TODO 工具只接受完整快照，防止增量补丁让模型与实际计划状态漂移。
export type TodoStatus = (typeof TODO_STATUSES)[number];

// 单项先 trim 再校验，并拒绝未知字段，避免脏数据进入会话计划。
const todoItemSchema = z
  .object({
    content: z
      .string()
      .transform((content) => content.trim())
      .pipe(z.string().min(1, "todo content must not be empty")),
    status: z.enum(TODO_STATUSES),
  })
  .strict();

// 工具参数始终是完整快照，不做增量补丁，保持模型看到的计划与内部状态一致。
const todoWriteSchema = z
  .object({
    todos: z.array(todoItemSchema).max(MAX_TODOS),
  })
  .strict();

export type TodoItem = Readonly<z.output<typeof todoItemSchema>>;
type TodoWriteInput = z.output<typeof todoWriteSchema>;

// TodoTracker 是会话级状态：每次构建 Agent 都新建实例，避免跨会话共享计划。
export class TodoTracker {
  // 观察每轮工具调用；计划长期未更新时，在下次模型请求前注入提醒。
  #todos: readonly TodoItem[] = Object.freeze([]);
  #nonTodoToolRounds = 0;
  readonly toolDefinition: ToolDefinition<TodoWriteInput>;

  constructor() {
    // toolDefinition 与当前 tracker 绑定，作为工具和观察器共享同一份会话状态。
    this.toolDefinition = Object.freeze({
      name: "todo_write",
      description: "Replace the current TODO list with a complete task snapshot.",
      inputSchema: todoWriteSchema,
      effect: "write",
      handler: (input: TodoWriteInput) => this.#writeTodos(input),
    });
  }

  get todos(): readonly TodoItem[] {
    // 只读暴露当前快照，供测试和上层检查，不开放外部修改入口。
    return this.#todos;
  }

  recordToolRound(toolNames: readonly string[]): void {
    // 一轮中出现 todo_write 即视为新快照，重置陈旧计数器。
    if (toolNames.length === 0) {
      return;
    }
    if (toolNames.includes(this.toolDefinition.name)) {
      this.#nonTodoToolRounds = 0;
      return;
    }
    this.#nonTodoToolRounds += 1;
  }

  beforeModel(): readonly ChatMessage[] {
    if (this.#nonTodoToolRounds < STALE_TOOL_ROUNDS) {
      return [];
    }
    // 提醒是请求级上下文；立即清零可避免后续每轮重复注入。
    this.#nonTodoToolRounds = 0;
    return Object.freeze([systemMessage(TODO_STALE_REMINDER)]);
  }

  #writeTodos(input: TodoWriteInput): ToolResult {
    // 先由 Zod 完整校验，再一次替换快照；失败路径不会触碰旧状态。
    this.#todos = Object.freeze(
      input.todos.map((item) => Object.freeze({ content: item.content, status: item.status })),
    );
    this.#nonTodoToolRounds = 0;
    return toolSuccess(serializeSnapshot(this.#todos));
  }
}

function serializeSnapshot(todos: readonly TodoItem[]): string {
  // 使用确定性 JSON 作为模型可见状态，方便下一轮完整替换。
  const json = JSON.stringify({
    todos: todos.map((item) => ({ content: item.content, status: item.status })),
  });
  const ascii: string[] = [];

  for (const character of json) {
    const codePoint = character.codePointAt(0);
    if (codePoint === undefined) {
      throw new Error("todo snapshot contained an invalid Unicode value");
    }

    if (codePoint <= 0x7f) {
      ascii.push(character);
      continue;
    }

    if (codePoint <= 0xffff) {
      ascii.push(`\\u${codePoint.toString(16).padStart(4, "0")}`);
      continue;
    }

    const offset = codePoint - 0x10000;
    const high = 0xd800 + (offset >> 10);
    const low = 0xdc00 + (offset & 0x3ff);
    ascii.push(`\\u${high.toString(16)}\\u${low.toString(16)}`);
  }

  return ascii.join("");
}
